src/main/java/bio/terra/workspace/service/resource/controlled/ControlledAiNotebookInstanceResource.java

```diff
@@ -79,6 +79,13 @@ public String attributesToJson() {
         new ControlledAiNotebookInstanceAttributes(getInstanceId(), getLocation()));
   }
 
+  @Override
+  public boolean matchesUniqueAttributes(String attributesJson) {
+    final ControlledAiNotebookInstanceAttributes attributes =
+        DbSerDes.fromJson(attributesJson, ControlledAiNotebookInstanceAttributes.class);
+    return getInstanceId().equals(attributes.getInstanceId());
```

Comment:
Is this unique enough without the location too?

Reply:
No, it's unique by (projectId, location, instanceId).

Reply:
Thanks. 

Unfortunately, I don't see an easy way to get the projectId  from here and the projectID isn't in the attributes structure, so my scheme only generalizes to one controlled resource type.

Reply:
Workspace ID would be a pretty good proxy for project ID I would think, if that helps. I don't like adding stuff to the attributes structure just to make this work, especially since every GCP resource will have a project ID.

Reply:
I think workspace id as a proxy for the project id is fine with a comment about what's going on. But it's also not crazy to add the project id to the attributes - it is an attribute to identify the notebook instance that isn't going anywhere and its frequently useful. But it does duplicate information.
Every GCP resource doesn't need to have the project id, just most - buckets are globally namespaced.

Reply:
If I leave it as just a check on instance ID and location, the risk is that we fail to allow a duplicate ID in another project when that should otherwise be allowed. It wouldn't be obvious to the user what happened. Which means maybe it's a lesser evil to just return false unconditionally for AI notebooks for now until we do the uri column or something similar.

Reply:
Why don't we do the right thing now? We can pass the workspace id or even the whole existing ControlledResource to this functions instead of just the attributes text. I'm not seeing the value in having an incorrect solution merged to dev.
If this always returns false, we're allowing controlled resource entries to refer to the same resource. Then we have to protect against granting inappropriate access on "private" notebook instances that are controlled by different users.

Reply:
I think we should give up on making this generic. For most resources, we only care about collisions within the WSM project. But buckets are in a global namespace and need the full table scan.

We already have a switch statement to put the right steps into the create, so I don't think we are making things particularly worse by having a separate checking step.

If most of the steps turn out to be resource specific, we can rethink the single create flight.